File: 18-Binary-Tree-Part-II/python/05_lowest_common_ancestor_of_a_binary_tree.py
from collections import deque
from typing import Optional

# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution:
    def lowestCommonAncestor(self, root: 'TreeNode', p: 'TreeNode', q: 'TreeNode') -> 'TreeNode':
        parent = {root: None}
        queue = deque([root])

        while p not in parent or q not in parent:
            node = queue.popleft()
            if node.left:
                parent[node.left] = node
                queue.append(node.left)
            if node.right:
                parent[node.right] = node
                queue.append(node.right)

        ancestors = set()
        current = p
        while current:
            ancestors.add(current)
            current = parent[current]

        current = q
        while current not in ancestors:
            current = parent[current]

        return current

File: 18-Binary-Tree-Part-II/cpp/05-lowest-common-ancestor-of-a-binary-tree.cpp
#include <unordered_map>
#include <unordered_set>
#include <queue>

using namespace std;

/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    TreeNode* lowestCommonAncestor(TreeNode* root, TreeNode* p, TreeNode* q) {
        unordered_map<TreeNode*, TreeNode*> parent;
        queue<TreeNode*> queue;
        parent[root] = nullptr;
        queue.push(root);

        while (parent.find(p) == parent.end() || parent.find(q) == parent.end()) {
            TreeNode* node = queue.front();
            queue.pop();
            if (node->left) {
                parent[node->left] = node;
                queue.push(node->left);
            }
            if (node->right) {
                parent[node->right] = node;
                queue.push(node->right);
            }
        }

        unordered_set<TreeNode*> ancestors;
        TreeNode* current = p;
        while (current != nullptr) {
            ancestors.insert(current);
            current = parent[current];
        }

        current = q;
        while (ancestors.find(current) == ancestors.end()) {
            current = parent[current];
        }

        return current;
    }
};
